Move BitmapImage metadata retrieving to a new class named BitmapImageDescriptor

shallawa opened the pull request:
#### b6e2e8c138383ce2f31e2c3b550246835682b577
<pre>
Move BitmapImage metadata retrieving to a new class named BitmapImageDescriptor
<a href="https://bugs.webkit.org/show_bug.cgi?id=274716">https://bugs.webkit.org/show_bug.cgi?id=274716</a>
<a href="https://rdar.apple.com/128738620">rdar://128738620</a>

Reviewed by Cameron McCormack.

This will decouple retrieving and caching the BitmapImage metadata from the
heavy weight class BitmapImageSource.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp: Added.
(WebCore::BitmapImageDescriptor::BitmapImageDescriptor):
(WebCore::BitmapImageDescriptor::imageMetadata const):
(WebCore::BitmapImageDescriptor::primaryImageFrameMetadata const):
(WebCore::BitmapImageDescriptor::primaryNativeImageMetadata const):
(WebCore::BitmapImageDescriptor::encodedDataStatus const):
(WebCore::BitmapImageDescriptor::size const):
(WebCore::BitmapImageDescriptor::sourceSize const):
(WebCore::BitmapImageDescriptor::densityCorrectedSize const):
(WebCore::BitmapImageDescriptor::orientation const):
(WebCore::BitmapImageDescriptor::primaryFrameIndex const):
(WebCore::BitmapImageDescriptor::frameCount const):
(WebCore::BitmapImageDescriptor::repetitionCount const):
(WebCore::BitmapImageDescriptor::colorSpace const):
(WebCore::BitmapImageDescriptor::singlePixelSolidColor const):
(WebCore::BitmapImageDescriptor::uti const):
(WebCore::BitmapImageDescriptor::filenameExtension const):
(WebCore::BitmapImageDescriptor::accessibilityDescription const):
(WebCore::BitmapImageDescriptor::hotSpot const):
(WebCore::BitmapImageDescriptor::maximumSubsamplingLevel const):
(WebCore::BitmapImageDescriptor::subsamplingLevelForScaleFactor const):
(WebCore::BitmapImageDescriptor::shouldUseQuickLookForFullscreen const):
(WebCore::BitmapImageDescriptor::dump const):
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h: Added.
(WebCore::BitmapImageDescriptor::clear):
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::BitmapImageSource):
(WebCore::BitmapImageSource::dataChanged):
(WebCore::BitmapImageSource::dump const):
(WebCore::BitmapImageSource::imageMetadata const): Deleted.
(WebCore::BitmapImageSource::primaryNativeImageMetadata const): Deleted.
(WebCore::BitmapImageSource::primaryImageFrameMetadata const): Deleted.
(WebCore::BitmapImageSource::encodedDataStatus const): Deleted.
(WebCore::BitmapImageSource::size const): Deleted.
(WebCore::BitmapImageSource::sourceSize const): Deleted.
(WebCore::BitmapImageSource::densityCorrectedSize const): Deleted.
(WebCore::BitmapImageSource::orientation const): Deleted.
(WebCore::BitmapImageSource::primaryFrameIndex const): Deleted.
(WebCore::BitmapImageSource::frameCount const): Deleted.
(WebCore::BitmapImageSource::repetitionCount const): Deleted.
(WebCore::BitmapImageSource::colorSpace const): Deleted.
(WebCore::BitmapImageSource::singlePixelSolidColor const): Deleted.
(WebCore::BitmapImageSource::uti const): Deleted.
(WebCore::BitmapImageSource::filenameExtension const): Deleted.
(WebCore::BitmapImageSource::accessibilityDescription const): Deleted.
(WebCore::BitmapImageSource::hotSpot const): Deleted.
(WebCore::BitmapImageSource::maximumSubsamplingLevel const): Deleted.
(WebCore::BitmapImageSource::subsamplingLevelForScaleFactor): Deleted.
(WebCore::BitmapImageSource::shouldUseQuickLookForFullscreen const): Deleted.
* Source/WebCore/platform/graphics/BitmapImageSource.h:
(WebCore::BitmapImageSource::decoderIfExists const):
(WebCore::BitmapImageSource::frames const):
(WebCore::BitmapImageSource::repetitionCount const):
(WebCore::BitmapImageSource::encodedDataStatus const):
(WebCore::BitmapImageSource::densityCorrectedSize const):
(WebCore::BitmapImageSource::maximumSubsamplingLevel const):
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:

Canonical link: <a href="https://commits.webkit.org/279710@main">https://commits.webkit.org/279710@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f0af231fd73220f3f63449b124f2ad8bd7dc7d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54231 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33620 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6773 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57508 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4957 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56534 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/41189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4904 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43931 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3323 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56327 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/41189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46970 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25073 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/41189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4297 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3104 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/41189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4504 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59100 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/4654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51349 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30609 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47079 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50711 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11818 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30392 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
```diff
@@ -0,0 +1,118 @@
+    mutable std::optional<IntPoint> m_hotSpot;
+    mutable SubsamplingLevel m_maximumSubsamplingLevel { SubsamplingLevel::Default };
+
+    BitmapImageSource& m_source;
```
Per https://a1391192.slack.com/archives/GK0ATJ6FJ/p1707106610975729 can you make this a ThreadSafeWeakPtr?

shallawa replied:
This caused a MM regression in the Leaves sub-test. So I had to take it out. Maybe after the recent changes in `BitmapImage` class and its friends, this is not true anymore.

So I will address this in a separate a separate patch after making sure it does not cause any regression.

shallawa committed:
Move BitmapImage metadata retrieving to a new class named BitmapImageDescriptor
https://bugs.webkit.org/show_bug.cgi?id=274716
rdar://128738620

Reviewed by Cameron McCormack.

This will decouple retrieving and caching the BitmapImage metadata from the
heavy weight class BitmapImageSource.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp: Added.
(WebCore::BitmapImageDescriptor::BitmapImageDescriptor):
(WebCore::BitmapImageDescriptor::imageMetadata const):
(WebCore::BitmapImageDescriptor::primaryImageFrameMetadata const):
(WebCore::BitmapImageDescriptor::primaryNativeImageMetadata const):
(WebCore::BitmapImageDescriptor::encodedDataStatus const):
(WebCore::BitmapImageDescriptor::size const):
(WebCore::BitmapImageDescriptor::sourceSize const):
(WebCore::BitmapImageDescriptor::densityCorrectedSize const):
(WebCore::BitmapImageDescriptor::orientation const):
(WebCore::BitmapImageDescriptor::primaryFrameIndex const):
(WebCore::BitmapImageDescriptor::frameCount const):
(WebCore::BitmapImageDescriptor::repetitionCount const):
(WebCore::BitmapImageDescriptor::colorSpace const):
(WebCore::BitmapImageDescriptor::singlePixelSolidColor const):
(WebCore::BitmapImageDescriptor::uti const):
(WebCore::BitmapImageDescriptor::filenameExtension const):
(WebCore::BitmapImageDescriptor::accessibilityDescription const):
(WebCore::BitmapImageDescriptor::hotSpot const):
(WebCore::BitmapImageDescriptor::maximumSubsamplingLevel const):
(WebCore::BitmapImageDescriptor::subsamplingLevelForScaleFactor const):
(WebCore::BitmapImageDescriptor::shouldUseQuickLookForFullscreen const):
(WebCore::BitmapImageDescriptor::dump const):
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h: Added.
(WebCore::BitmapImageDescriptor::clear):
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::BitmapImageSource):
(WebCore::BitmapImageSource::dataChanged):
(WebCore::BitmapImageSource::dump const):
(WebCore::BitmapImageSource::imageMetadata const): Deleted.
(WebCore::BitmapImageSource::primaryNativeImageMetadata const): Deleted.
(WebCore::BitmapImageSource::primaryImageFrameMetadata const): Deleted.
(WebCore::BitmapImageSource::encodedDataStatus const): Deleted.
(WebCore::BitmapImageSource::size const): Deleted.
(WebCore::BitmapImageSource::sourceSize const): Deleted.
(WebCore::BitmapImageSource::densityCorrectedSize const): Deleted.
(WebCore::BitmapImageSource::orientation const): Deleted.
(WebCore::BitmapImageSource::primaryFrameIndex const): Deleted.
(WebCore::BitmapImageSource::frameCount const): Deleted.
(WebCore::BitmapImageSource::repetitionCount const): Deleted.
(WebCore::BitmapImageSource::colorSpace const): Deleted.
(WebCore::BitmapImageSource::singlePixelSolidColor const): Deleted.
(WebCore::BitmapImageSource::uti const): Deleted.
(WebCore::BitmapImageSource::filenameExtension const): Deleted.
(WebCore::BitmapImageSource::accessibilityDescription const): Deleted.
(WebCore::BitmapImageSource::hotSpot const): Deleted.
(WebCore::BitmapImageSource::maximumSubsamplingLevel const): Deleted.
(WebCore::BitmapImageSource::subsamplingLevelForScaleFactor): Deleted.
(WebCore::BitmapImageSource::shouldUseQuickLookForFullscreen const): Deleted.
* Source/WebCore/platform/graphics/BitmapImageSource.h:
(WebCore::BitmapImageSource::decoderIfExists const):
(WebCore::BitmapImageSource::frames const):
(WebCore::BitmapImageSource::repetitionCount const):
(WebCore::BitmapImageSource::encodedDataStatus const):
(WebCore::BitmapImageSource::densityCorrectedSize const):
(WebCore::BitmapImageSource::maximumSubsamplingLevel const):
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:

Canonical link: https://commits.webkit.org/279710@main
